From dc463a94a8e3d1bcb864797a4f3aab65bab41e4c Mon Sep 17 00:00:00 2001
From: Rahul Arya <aryarahul@google.com>
Date: Wed, 14 Sep 2022 19:02:37 +0000
Subject: [PATCH] Reschedule address rotation after connection to advertising
 set

Otherwise we stop rotating our address and lose all benefits of privacy.

Ignore-AOSP-First: Security fix
Bug: 246692649
Bug: 195410559
Test: Nearby QA test, and I'll write a cert test eventually
Tag: #stability
Change-Id: Ib63d79256acbb956413849c17f761aa902fa2eae
(cherry picked from commit 279e3c3a8656808582e3ebcf81d8a936c320498d)
Merged-In: Ib63d79256acbb956413849c17f761aa902fa2eae
(cherry picked from commit acffc03f14b37144a5687525440effc8c6dd1b93)
Merged-In: Ib63d79256acbb956413849c17f761aa902fa2eae
---
 system/gd/hci/le_advertising_manager.cc | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/system/gd/hci/le_advertising_manager.cc b/system/gd/hci/le_advertising_manager.cc
index 4f744ed179..eca691b031 100644
--- a/system/gd/hci/le_advertising_manager.cc
+++ b/system/gd/hci/le_advertising_manager.cc
@@ -182,7 +182,9 @@ struct LeAdvertisingManager::impl : public bluetooth::hci::LeAddressManagerCallb
 
     uint8_t advertiser_id = event_view.GetAdvertisingHandle();
 
+    bool was_rotating_address = false;
     if (advertising_sets_[advertiser_id].address_rotation_alarm != nullptr) {
+      was_rotating_address = true;
       advertising_sets_[advertiser_id].address_rotation_alarm->Cancel();
       advertising_sets_[advertiser_id].address_rotation_alarm.reset();
     }
@@ -209,6 +211,13 @@ struct LeAdvertisingManager::impl : public bluetooth::hci::LeAddressManagerCallb
       if (advertising_sets_[advertiser_id].duration == 0 &&
           advertising_sets_[advertiser_id].max_extended_advertising_events == 0) {
         LOG_INFO("Reenable advertising");
+        if (was_rotating_address) {
+          advertising_sets_[advertiser_id].address_rotation_alarm = std::make_unique<os::Alarm>(module_handler_);
+          advertising_sets_[advertiser_id].address_rotation_alarm->Schedule(
+              common::BindOnce(
+                  &impl::set_advertising_set_random_address_on_timer, common::Unretained(this), advertiser_id),
+              le_address_manager_->GetNextPrivateAddressIntervalMs());
+        }
         enable_advertiser(advertiser_id, true, 0, 0);
       }
     }
-- 
2.38.1.273.g43a17bfeac-goog

